## Data Cleaning 

#### Table of Contents: 
#### 1. Importing Libraries and reading data
#### 2. Exploratory Data Analysis 
#### 3. Data Cleaning
#### 4. Checking Coverage Rate
#### 5. Summary and Conclusion

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

test_ad = pd.read_csv('./data/test.csv')
real_ad = pd.read_csv('./data/all_addresses.csv')

##### First, let's examine the full list of real addresses 

In [2]:
#see what first 5 rows look like
real_ad.head()

address           city state    zip
0      941 Thorpe St   Rock Springs    WY  82901
1     2809 HARRIS DR        Antioch    CA  94509
2  1441 Eastlake Ave    Los Angeles    CA  90033
3       7 Eucalyptus  Newport Beach    CA  92657
4    1400 Lachman Ln    Los Angeles    CA  90272

In [210]:
# rename two of the columns 
real_ad.rename(columns={'address':'street_nm', 'zip':'zip_code'}, inplace=True)
real_ad.columns

Index(['street_nm', 'city', 'state', 'zip_code', 'full_address'], dtype='object')

In [212]:
real_ad.columns

Index(['street_nm', 'city', 'state', 'zip_code', 'full_address'], dtype='object')

In [213]:
real_ad.dtypes

street_nm       object
city            object
state           object
zip_code        object
full_address    object
dtype: object

In [244]:
res = test_ad[~test_ad.address.isin(real_ad.street_nm)]

In [242]:
res.shape

(27153, 7)

In [246]:
res.sample(20)

uid                            address                   city state  \
10468  10548                          15103rdSt        Manhattan Beach    CA   
20811  20961     224 W Sherwood Blvd, Apt 20812               Big Bear    CA   
53727  54120     37319 Paseo Tulipa, Unit 53728               Murrieta    CA   
40569  40880  19380 Country View Dr, Unit 40570             Cottonwood    CA   
11204  11288                 360 Dxie Canyon Pl            Los Angeles    CA   
90322  91010       605 S Myrtle Ave, Unit 90323               Monrovia    CA   
6032    6084                        17610HadaDr              San Diego    CA   
33413  33652                     683 beverly pl             san marcos    ca   
7652    7712                     860 NGlenn Ave                 Fresno    CA   
37267  37540      363 Lakeover Dr W, Unit 37268               Columbus    MS   
90740  91430      8040 Gilleland Dr, Unit 90741              Roseville    CA   
78863  79458                  2766E Vn Buren St             Long Beach    CA   
59307  59748                      1415 beech st         south pasadena    ca   
23002  23164                      5651 cedar st              riverside    ca   
17496  17628                          2 GachoDr  Rolling Hills Estates    CA   
56825  57244                    1537 addison rd   palos verdes estates    ca   
76024  76590       7409 Chilibre St, Unit 76025            Bakersfield    CA   
7118    7175           18055 Coastline Dr #7119                 Malibu    CA   
24099  24273           123 Overbrook, Apt 24100                 Irvine    CA   
3099    3124                         81 Caiente                Barstow    CA   

         zip                                       Cust_address  \
10468  90266                  15103rdStManhattan BeachCA90266.0   
20811  92314    224 W Sherwood Blvd, Apt 20812Big BearCA92314.0   
53727  92563    37319 Paseo Tulipa, Unit 53728MurrietaCA92563.0   
40569  96022  19380 Country View Dr, Unit 40570CottonwoodCA9...   
11204  91423             360 Dxie Canyon PlLos AngelesCA91423.0   
90322  91016      605 S Myrtle Ave, Unit 90323MonroviaCA91016.0   
6032   92127                      17610HadaDrSan DiegoCA92127.0   
33413  92078                  683 beverly plsan marcosca92078.0   
7652   93711                      860 NGlenn AveFresnoCA93711.0   
37267  39702     363 Lakeover Dr W, Unit 37268ColumbusMS39702.0   
90740  95747    8040 Gilleland Dr, Unit 90741RosevilleCA95747.0   
78863  90810               2766E Vn Buren StLong BeachCA90810.0   
59307  91030               1415 beech stsouth pasadenaca91030.0   
23002  92509                    5651 cedar striversideca92509.0   
17496  90274            2 GachoDrRolling Hills EstatesCA90274.0   
56825  90274       1537 addison rdpalos verdes estatesca90274.0   
76024  93313   7409 Chilibre St, Unit 76025BakersfieldCA93313.0   
7118   90265            18055 Coastline Dr #7119MalibuCA90265.0   
24099  92620            123 Overbrook, Apt 24100IrvineCA92620.0   
3099   92311                         81 CaienteBarstowCA92311.0   

                                            full_address  
10468                    15103rdStManhattan BeachCA90266  
20811      224 W Sherwood Blvd, Apt 20812Big BearCA92314  
53727      37319 Paseo Tulipa, Unit 53728MurrietaCA92563  
40569  19380 Country View Dr, Unit 40570CottonwoodCA9...  
11204               360 Dxie Canyon PlLos AngelesCA91423  
90322        605 S Myrtle Ave, Unit 90323MonroviaCA91016  
6032                         17610HadaDrSan DiegoCA92127  
33413                    683 beverly plsan marcosca92078  
7652                         860 NGlenn AveFresnoCA93711  
37267       363 Lakeover Dr W, Unit 37268ColumbusMS39702  
90740      8040 Gilleland Dr, Unit 90741RosevilleCA95747  
78863                 2766E Vn Buren StLong BeachCA90810  
59307                 1415 beech stsouth pasadenaca91030  
23002                      5651 cedar striversideca92509  
17496              2 GachoDrRolling

In [225]:
len(res)

27153

In [226]:
res2 = test_ad.address[test_ad.address.isin(real_ad.street_nm)].values
len(res2)

72096

In [239]:
#compare values in each row to values in corresponding columns 
initial_match =  test_ad[(test_ad.address.isin(real_ad.street_nm)) & (test_ad.city.isin(real_ad.city))
                                 & (test_ad.state.isin(real_ad.state)) & (test_ad.zip.isin(real_ad.zip_code))]


In [240]:
match_pct = (len(initial_match) * 100) /len(real_ad)
match_pct

49.635384615384616

In [294]:
match_pct = (len(initial_match) * 100) /len(real_ad)
print("The test dataset has a {}% match".format(match_pct)) 

The test dataset has a 49.635384615384616% match


In [219]:
cols = ['street_nm', 'city', 'state', 'zip_code', 'full_address']
real_ad[cols] = real_ad[cols].apply(lambda x: x.str.strip())

In [220]:
real_ad['full_address'] = real_ad['street_nm'] + real_ad['city'] + real_ad['state']+ real_ad['zip_code']

In [205]:
real_ad.head(3)

street_nm          city state zip_code  \
0      941 Thorpe St  Rock Springs    WY    82901   
1     2809 HARRIS DR       Antioch    CA    94509   
2  1441 Eastlake Ave   Los Angeles    CA    90033   

                          full_address  
0     941 Thorpe StRock SpringsWY82901  
1         2809 HARRIS DRAntiochCA94509  
2  1441 Eastlake AveLos AngelesCA90033

In [247]:
len(test_ad)/len(real_ad)

0.7634538461538461

In [208]:
real_ad.full_address.dtypes

dtype('O')

In [206]:
real_ad.dtypes

street_nm       object
city            object
state           object
zip_code        object
full_address    object
dtype: object

In [78]:
test_ad['Cust_address'] = test_ad['address'] + test_ad['city'] + test_ad['state']+ test_ad['zip'].astype(str)

In [42]:
ad_lst = real_ad.full_address.tolist()

In [72]:
test_ad.head()

uid            address           city state      zip  \
0    1      941 Thorpe St   Rock Springs    WY  82901.0   
1    2     2809 HARRIS DR        Antioch    CA  94509.0   
2    3  1441 Eastlake Ave    Los Angeles    CA  90033.0   
3    4         7 ucayptus  Newport Beach    CA  92657.0   
4    5    1400 Lachman Ln            NaN    CA  90272.0   

                            Cust_address  
0     941 Thorpe StRock SpringsWY82901.0  
1         2809 HARRIS DRAntiochCA94509.0  
2  1441 Eastlake AveLos AngelesCA90033.0  
3       7 ucayptusNewport BeachCA92657.0  
4                                    NaN

In [131]:
real_ad.head()

address           city state    zip  \
0      941 Thorpe St   Rock Springs    WY  82901   
1     2809 HARRIS DR        Antioch    CA  94509   
2  1441 Eastlake Ave    Los Angeles    CA  90033   
3       7 Eucalyptus  Newport Beach    CA  92657   
4    1400 Lachman Ln    Los Angeles    CA  90272   

                          full_address  
0     941 Thorpe StRock SpringsWY82901  
1         2809 HARRIS DRAntiochCA94509  
2  1441 Eastlake AveLos AngelesCA90033  
3     7 EucalyptusNewport BeachCA92657  
4    1400 Lachman LnLos AngelesCA90272

In [295]:
print("The dataset has the following shape: {}.".format(real_ad.shape))

The dataset has the following shape: (130000, 5).


In [141]:
real_ad.groupby('full_address').size()

full_address
#NAME?---                                 1
#NAME?-CA-                               27
#NAME?-CA91390                            5
#NAME?-CA91901                            2
#NAME?-CA92036                            2
                                         ..
Yaqui Pass RdBorrego SpringsCA92004       4
Yerington AveKeelerCA93530                1
Zephyr RdHelendaleCA92342                 1
Zinfandel DrRancho CordovaCA95670         1
undefined Marines DrLittle ElmTX75068     1
Length: 96632, dtype: int64

In [320]:
#examine columns that start with ' - ' or ' # '
dirty = real_ad[real_ad['full_address'].astype(str).str.startswith('#') | real_ad['full_address'].astype(str).str.startswith('-')| 
                real_ad['city'].astype(str).str.startswith('-') | real_ad['state'].astype(str).str.startswith('-') ]
dirty.shape

(1055, 5)

In [ ]:
def clean_addr()

In [316]:
 real_ad[real_ad['full_address'].astype(str).str.startswith('-')].head()

street_nm          city state zip_code              full_address
529   - E Ave R 10             -    CA    93543      - E Ave R 10-CA93543
2091     - 17th St      Piedmont    CA    94611  - 17th StPiedmontCA94611
2618           - -  Butte Valley    CA    95965    - -Butte ValleyCA95965
2986           - -      Palmdale    CA    93552        - -PalmdaleCA93552
3037           - -     Lancaster    CA    93536       - -LancasterCA93536

In [319]:
 real_ad[real_ad['zip_code'].astype(str).str.startswith('-')].head()

street_nm       city state zip_code             full_address
106         #NAME?    Hayward    CA        -         #NAME?HaywardCA-
3569        #NAME?          -    CA        -               #NAME?-CA-
4511        #NAME?          -    CA        -               #NAME?-CA-
10102  - 40th St E  Lancaster    CA        -  - 40th St ELancasterCA-
10799       #NAME?          -    CA        -               #NAME?-CA-

In [314]:
clean = real_ad[~real_ad['full_address'].astype(str).str.startswith('#') | real_ad['full_address'].astype(str).str.startswith('-')]
clean.shape

(129496, 5)

In [154]:
clean.head()

address           city state    zip  \
0      941 Thorpe St   Rock Springs    WY  82901   
1     2809 HARRIS DR        Antioch    CA  94509   
2  1441 Eastlake Ave    Los Angeles    CA  90033   
3       7 Eucalyptus  Newport Beach    CA  92657   
4    1400 Lachman Ln    Los Angeles    CA  90272   

                          full_address  
0     941 Thorpe StRock SpringsWY82901  
1         2809 HARRIS DRAntiochCA94509  
2  1441 Eastlake AveLos AngelesCA90033  
3     7 EucalyptusNewport BeachCA92657  
4    1400 Lachman LnLos AngelesCA90272

In [298]:
clean.full_address.dtypes

dtype('O')

In [300]:
addr = clean['full_address'].astype(str)
addr.head()

0       941 Thorpe StRock SpringsWY82901
1           2809 HARRIS DRAntiochCA94509
2    1441 Eastlake AveLos AngelesCA90033
3       7 EucalyptusNewport BeachCA92657
4      1400 Lachman LnLos AngelesCA90272
Name: full_address, dtype: object

In [321]:
#addr.to_list()

In [305]:
addr.str.strip()

0            941 Thorpe StRock SpringsWY82901
1                2809 HARRIS DRAntiochCA94509
2         1441 Eastlake AveLos AngelesCA90033
3            7 EucalyptusNewport BeachCA92657
4           1400 Lachman LnLos AngelesCA90272
                         ...                 
129995       4477 Camrose AveSan DiegoCA92122
129996              13 STONEWALLIrvineCA92620
129997      3456 Nouveau WayGold RiverCA95670
129998      9115 Bungalow WayElk GroveCA95758
129999    1824 Highlands View RdAlpineCA91901
Name: full_address, Length: 129496, dtype: object

In [ ]:
bigdata = data1.append(data2, ignore_index=True)

In [158]:
common_ad = real_ad.append(test_ad, ignore_index=True, sort=True)
common_ad.shape

(229249, 7)

In [159]:
common_ad.head()

Cust_address            address           city  \
0          NaN      941 Thorpe St   Rock Springs   
1          NaN     2809 HARRIS DR        Antioch   
2          NaN  1441 Eastlake Ave    Los Angeles   
3          NaN       7 Eucalyptus  Newport Beach   
4          NaN    1400 Lachman Ln    Los Angeles   

                          full_address state  uid    zip  
0     941 Thorpe StRock SpringsWY82901    WY  NaN  82901  
1         2809 HARRIS DRAntiochCA94509    CA  NaN  94509  
2  1441 Eastlake AveLos AngelesCA90033    CA  NaN  90033  
3     7 EucalyptusNewport BeachCA92657    CA  NaN  92657  
4    1400 Lachman LnLos AngelesCA90272    CA  NaN  90272

In [163]:
new_df = common_ad.drop_duplicates(subset=['address', 'full_address'])
new_df.shape

(193754, 7)

In [81]:
df1 = real_ad[['full_address']]

In [82]:
df2 = test_ad[['Cust_address']]

In [84]:
df3 = pd.concat([df1,df2 ], axis=1)
df3.head()

full_address                           Cust_address
0     941 Thorpe StRock SpringsWY82901     941 Thorpe StRock SpringsWY82901.0
1         2809 HARRIS DRAntiochCA94509         2809 HARRIS DRAntiochCA94509.0
2  1441 Eastlake AveLos AngelesCA90033  1441 Eastlake AveLos AngelesCA90033.0
3     7 EucalyptusNewport BeachCA92657       7 ucayptusNewport BeachCA92657.0
4    1400 Lachman LnLos AngelesCA90272                                    NaN

In [85]:
df3.shape

(130000, 2)

In [90]:
df3['Cust_address'] = df3['Cust_address'].astype(str).apply(lambda x: x.replace('.0',''))
df3.head(3)

full_address                         Cust_address  \
0     941 Thorpe StRock SpringsWY82901     941 Thorpe StRock SpringsWY82901   
1         2809 HARRIS DRAntiochCA94509         2809 HARRIS DRAntiochCA94509   
2  1441 Eastlake AveLos AngelesCA90033  1441 Eastlake AveLos AngelesCA90033   

   match  
0      0  
1      0  
2      0

In [91]:
df3['match'] = np.where(df3.full_address == df3.Cust_address,1,0)
df3.head()

full_address                         Cust_address  \
0     941 Thorpe StRock SpringsWY82901     941 Thorpe StRock SpringsWY82901   
1         2809 HARRIS DRAntiochCA94509         2809 HARRIS DRAntiochCA94509   
2  1441 Eastlake AveLos AngelesCA90033  1441 Eastlake AveLos AngelesCA90033   
3     7 EucalyptusNewport BeachCA92657       7 ucayptusNewport BeachCA92657   
4    1400 Lachman LnLos AngelesCA90272                                  nan   

   match  
0      1  
1      1  
2      1  
3      0  
4      0

In [92]:
df3.groupby('match').size()

match
0    129983
1        17
dtype: int64

In [100]:
sum(df3.match) * 100 /len(df3.match)

0.013076923076923076

In [102]:
#examine rows that did not match

df3[df3.match == 0]

full_address  \
3          7 EucalyptusNewport BeachCA92657   
4         1400 Lachman LnLos AngelesCA90272   
5       725 Mountain View StAltadenaCA91001   
7             4341 69th StSacramentoCA95820   
9           3640 Oak Creek DrOntarioCA91761   
...                                     ...   
129995     4477 Camrose AveSan DiegoCA92122   
129996            13 STONEWALLIrvineCA92620   
129997    3456 Nouveau WayGold RiverCA95670   
129998    9115 Bungalow WayElk GroveCA95758   
129999  1824 Highlands View RdAlpineCA91901   

                                    Cust_address  match  
3                 7 ucayptusNewport BeachCA92657      0  
4                                            nan      0  
5              725 ounain View StAltadenaCA91001      0  
7                    431 6th StSacramentoCA95820      0  
9       3640 Oak Creek Dr, Unit 10OntarioCA91761      0  
...                                          ...    ...  
129995                                       nan      0  
129996                                       nan      0  
129997                                       nan      0  
129998                                       nan      0  
129999                                       nan      0  

[129983 rows x 3 columns]

In [107]:
df3['full_address'] = df3.full_address.str.lower()
df3['Cust_address'] = df3.full_address.str.lower()
df3.head()

full_address                         Cust_address  \
0     941 thorpe strock springswy82901     941 thorpe strock springswy82901   
1         2809 harris drantiochca94509         2809 harris drantiochca94509   
2  1441 eastlake avelos angelesca90033  1441 eastlake avelos angelesca90033   
3     7 eucalyptusnewport beachca92657     7 eucalyptusnewport beachca92657   
4    1400 lachman lnlos angelesca90272    1400 lachman lnlos angelesca90272   

   match  
0      1  
1      1  
2      1  
3      0  
4      0

In [108]:
df3['match'] = np.where(df3.full_address == df3.Cust_address,1,0)
df3.head()

full_address                         Cust_address  \
0     941 thorpe strock springswy82901     941 thorpe strock springswy82901   
1         2809 harris drantiochca94509         2809 harris drantiochca94509   
2  1441 eastlake avelos angelesca90033  1441 eastlake avelos angelesca90033   
3     7 eucalyptusnewport beachca92657     7 eucalyptusnewport beachca92657   
4    1400 lachman lnlos angelesca90272    1400 lachman lnlos angelesca90272   

   match  
0      1  
1      1  
2      1  
3      1  
4      1

In [109]:
new_match = sum(df3.match) * 100 /len(df3.match)
new_match

99.77307692307693

In [112]:
df3[df3.match == 0].head()

full_address Cust_address  match
341           NaN          NaN      0
764           NaN          NaN      0
967           NaN          NaN      0
1034          NaN          NaN      0
1057          NaN          NaN      0

In [115]:
df4 = df3.full_address
df4.head()

0       941 thorpe strock springswy82901
1           2809 harris drantiochca94509
2    1441 eastlake avelos angelesca90033
3       7 eucalyptusnewport beachca92657
4      1400 lachman lnlos angelesca90272
Name: full_address, dtype: object

In [116]:
df5 = df3.Cust_address
df5.head()

0       941 thorpe strock springswy82901
1           2809 harris drantiochca94509
2    1441 eastlake avelos angelesca90033
3       7 eucalyptusnewport beachca92657
4      1400 lachman lnlos angelesca90272
Name: Cust_address, dtype: object

In [ ]:
df3['match'] = np.where(df3.full_address == df3.Cust_address,1,0)
df3.head()

In [122]:
df3['check'] = np.where(df3.Cust_address == df3.full_address,1,0)

In [123]:
df3.head(3)

full_address                         Cust_address  \
0     941 thorpe strock springswy82901     941 thorpe strock springswy82901   
1         2809 harris drantiochca94509         2809 harris drantiochca94509   
2  1441 eastlake avelos angelesca90033  1441 eastlake avelos angelesca90033   

   match  check  
0      1      1  
1      1      1  
2      1      1

In [129]:
df3.check.dtypes

dtype('int64')

In [126]:
ouput = df3.groupby('check').size()

In [130]:
print(sum(df3.check) * 100 / len(df3.check))

99.77307692307693


In [3]:
#count of rows and columns
real_ad.shape

(130000, 4)

In [119]:
df3.dtypes

full_address    object
Cust_address    object
match            int64
dtype: object

In [5]:
#info on dataset
real_ad.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130000 entries, 0 to 129999
Data columns (total 4 columns):
address    129705 non-null object
city       129984 non-null object
state      129984 non-null object
zip        129984 non-null object
dtypes: object(4)
memory usage: 4.0+ MB


In [6]:
#examine the address column - using sample
real_ad['address'].sample(10)

128395         39 Garrison Loop
90520            21623 Ulmus Dr
103935             - W Ave E 14
101322       30061 Crestview Ct
2295      2490 Bald Mountain Rd
100964     38301 Ocean Ridge Dr
88731      5027 Shannon View Rd
54907          445 Yosemite Ave
82156           1171 Legion Ave
111827       1607 Centennial Dr
Name: address, dtype: object

In [7]:
#descriptive info on addres column
real_ad.address.describe()

count     129705
unique     96279
top       #NAME?
freq         504
Name: address, dtype: object

In [8]:
#check top 10 commonly used addresses
real_ad.address.value_counts()[:10]

#NAME?                   504
- -                      167
5500 Grand Pacific Dr    126
5805 Armada Dr           100
Fairway Pl                91
3254 Avila Beach Dr       76
7210 Blue Heron Pl        73
3075 Carlsbad Blvd        69
8858 Lawrence Welk Dr     56
25382 Pappas Rd           47
Name: address, dtype: int64

In [9]:
#check using groupby
real_ad.groupby('address').size()

address
#NAME?                  504
- -                     167
- 10th Ave                2
- 110th St W              2
- 115th St E              2
                       ... 
Yaqui Pass Rd             4
Yerington Ave             1
Zephyr Rd                 1
Zinfandel Dr              1
undefined Marines Dr      1
Length: 96279, dtype: int64

In [256]:
real_ad.columns

Index(['street_nm', 'city', 'state', 'zip_code', 'full_address'], dtype='object')

#### Some Key Observations on the address column:

1. There are 504 records of incorrect inputs with the value 'NAME#'
2. There are 167 missing values, with the input '- -'
3. Several actual addresses with wrong inputs, starting with '-'

### Let's examine the City column

In [10]:
real_ad.city.sample(10)

70884         Elk Grove
107010          Vallejo
127102          Cypress
109082         Paradise
78833          El Monte
1224      Valley Center
9698          Calabasas
121096       Sacramento
97750       San Jacinto
47557        Long Beach
Name: city, dtype: object

In [11]:
#top 5 cities 

real_ad['city'].value_counts()[:5]

Los Angeles    8059
San Diego      3994
San Jose       2691
Sacramento     2455
Bakersfield    1843
Name: city, dtype: int64

In [12]:
#examine city using groupby

real_ad.groupby('city').size()

city
-               524
ACTON             2
ADELANTO          1
AGOURA HILLS      2
AHWAHNEE          1
               ... 
Zebulon           1
Zenia             1
Zephyr Cove       1
Zion              1
Zionsville        2
Length: 3799, dtype: int64

#### Key Observation on City Column
- 504 records have missing values, with input '-'

#### Let's examine the state and zip columns

In [13]:
real_ad.state.sample(10)

102464    CA
9815      CA
98082     CA
1242      CA
32556     CA
36374     CA
48774     CA
26562     CA
9538      CA
88603     TX
Name: state, dtype: object

In [14]:
#top 5 states
real_ad.state.value_counts()[:5]

CA    120208
TX      1251
IL       892
NY       875
WA       832
Name: state, dtype: int64

In [15]:
#Examine using groupby
real_ad.groupby('state').size()

state
-                1
AK               7
AL              82
AR              57
AZ             314
Auckland         1
CA          120208
CHIH             2
CHIS             1
CO             715
CT             329
CÓRDOBA          1
DC               2
DE              10
FL             292
GA             170
GAUTENG          2
HI               6
IA              75
ID              84
IL             892
IN             117
KS             111
KY              94
LA              65
MA             202
MD              42
ME              35
MI              86
MN             134
MO              90
MS             115
MT              56
NC             545
ND              40
NE              52
NH              29
NJ              98
NM              73
NV             146
NY             875
OH             469
OK              94
ON               1
OR             131
PA             425
RI              20
SC              52
SD              26
TN             140
TX            1251
UT              56
VA    

In [16]:
real_ad.state.nunique()

59

#### Key Observations on the state column
- Missing values/incorrect value, '-'
- Incorrect inputs with names, e.g. GUATENG, instead of abbreviations
- Wrong info/abbreviations used, e.g. ON 
- Total of 59 unique values for state, insteead of 50

In [17]:
# Examining the zip column
real_ad.zip.sample(10)

48391     94526
6605      95358
122406    92356
89930     94403
30887     93022
16041     95945
87171     95404
75934     94596
126179    95247
67535     92782
Name: zip, dtype: object

In [18]:
#top 5 most common zip codes
real_ad.zip.value_counts().head()

92008    456
93535    428
92026    411
95969    400
93536    388
Name: zip, dtype: int64

In [19]:
#examine using groupy
real_ad.groupby('zip').size()

zip
-        77
10003     1
10009     4
1001      1
10025     2
         ..
99507     1
99508     1
99517     1
99709     2
M9B       1
Length: 5685, dtype: int64

#### Some observations on zip codes
- Missing/null values where the input is '-'
- Some records with the number of values less than 5 
- Incorrect formats where input has a combination of letters and numbers e.g.M9B

### Now let's examine the test dataset

In [20]:
test_ad.head()

uid            address           city state      zip
0    1      941 Thorpe St   Rock Springs    WY  82901.0
1    2     2809 HARRIS DR        Antioch    CA  94509.0
2    3  1441 Eastlake Ave    Los Angeles    CA  90033.0
3    4         7 ucayptus  Newport Beach    CA  92657.0
4    5    1400 Lachman Ln            NaN    CA  90272.0

In [21]:
test_ad.shape

(99249, 5)

In [22]:
test_ad.dtypes

uid          int64
address     object
city        object
state       object
zip        float64
dtype: object

In [23]:
#examine the address column using groupby
test_ad.groupby('address').size()

address
#NAM?              4
#NAME?           206
#NAME? #11606      1
#NAME? #13850      1
#NAME? #18941      1
                ... 
w commons          1
walnut rd          1
wildomar           1
willis ave         1
x st               1
Length: 97121, dtype: int64

In [24]:
#check the top 10 most common addresses

test_ad.address.value_counts().head(10)

#NAME?                         206
5500 Grand Pacific Dr           69
Fairway Pl                      53
5805 Armada Dr                  50
3254 Avila Beach Dr             47
7210 Blue Heron Pl              43
3075 Carlsbad Blvd              40
8858 Lawrence Welk Dr           29
25382 Pappas Rd                 24
2600 Avenida Del Presidente     24
Name: address, dtype: int64

In [25]:
test_ad['address'].tail()

99244                  2020 s moreno dr
99245                10641 Missouri Ave
99246                    27905 Tyler Ln
99247                    6259 W 55th St
99248    44282 Compiegne Dr, Unit 99249
Name: address, dtype: object

In [26]:
#unique addreses
test_ad.address.unique()

array(['941 Thorpe St', '2809 HARRIS DR', '1441 Eastlake Ave', ...,
       '27905 Tyler Ln', '6259 W 55th St',
       '44282 Compiegne Dr, Unit 99249'], dtype=object)

In [27]:
#examine the city column
test_ad.groupby('city').size()

city
ACTON              2
ADELANTO           1
AGOURA HILLS       1
AHWAHNEE           1
AIRWAY HEIGHTS     1
                  ..
ypsilanti          1
yreka              4
yuba city         13
yucaipa            8
yucca valley       5
Length: 4768, dtype: int64

In [28]:
#top 5 cities
test_ad.city.value_counts().head()

Los Angeles    5594
San Diego      2833
San Jose       1889
Sacramento     1704
Bakersfield    1291
Name: city, dtype: int64

In [29]:
#counts of state
test_ad.state.value_counts().head()

CA    78313
ca     6622
TX      837
IL      604
WA      548
Name: state, dtype: int64

In [30]:
#examine the state column
test_ad.groupby('state').size()

state
AK        5
AL       53
AR       35
AZ      200
CA    78313
      ...  
ut        4
va        5
wa       42
wi        6
wy        4
Length: 96, dtype: int64

In [31]:
test_ad['state'].unique()

array(['WY', 'CA', 'NY', 'TX', 'TN', 'PA', 'MI', 'FL', 'GA', 'MO', 'AR',
       'CO', 'WA', 'OH', 'AZ', 'MA', 'IL', 'CT', 'SC', 'KY', 'NV', 'MN',
       'AL', 'NC', 'OK', 'WV', 'VA', 'UT', 'VT', 'ID', 'NJ', 'IN', 'MD',
       'SD', 'OR', 'NH', 'NM', 'IA', 'KS', 'LA', 'WI', 'ME', 'NE', 'MS',
       'MT', 'DE', 'ND', 'ca', 'ma', 'ia', 'nd', 'oh', 'id', 'il', 'HI',
       'tx', 'ut', 'pa', 'sd', 'ny', 'RI', 'co', 'wa', 'va', 'mn', 'az',
       'nc', 'tn', nan, 'ne', 'ky', 'nj', 'la', 'ct', 'ks', 'mo', 'md',
       'al', 'wi', 'mi', 'wy', 'AK', 'fl', 'in', 'DC', 'ri', 'or', 'nh',
       'hi', 'nm', 'mt', 'ms', 'ga', 'nv', 'ar', 'ok', 'me'], dtype=object)

In [32]:
test_ad['state'].nunique()

96

In [33]:
#examine the zip codes

test_ad.zip.sample(10)

30595    92009.0
3088     93247.0
5882     92887.0
60623    92128.0
90019    90813.0
70694    93285.0
1630     96161.0
99059    94587.0
83783    91744.0
11477    95687.0
Name: zip, dtype: float64

In [34]:
#using groupby
test_ad.groupby('zip').size()

zip
1001.0     1
1020.0     3
1027.0     1
1028.0     2
1035.0     1
          ..
99504.0    1
99507.0    1
99508.0    1
99517.0    1
99709.0    1
Length: 5419, dtype: int64

#### Key Observations in the test dataset 
- Incorrect formats, missing values as expected in both address ad=nd city columns
- state column a combination of upper case and lower cases

In [25]:
#examining specific Scenarios: #NAME?

real_ad[real_ad.address == '#NAME?'].sample(10)

address       city state    zip
68965   #NAME?     Folsom    CA  95630
122499  #NAME?    Modesto    CA  95350
55727   #NAME?          -    CA  93514
89602   #NAME?          -    CA  93544
119471  #NAME?          -    CA  93535
64133   #NAME?          -    CA  91901
48064   #NAME?  Crestline    CA  92325
20174   #NAME?          -    CA  93536
29059   #NAME?  Crestline    CA      -
969     #NAME?          -    CA  93535

In [261]:
new_dict = real_ad.to_dict()

In [271]:
real_ad.head(3)

street_nm          city state zip_code  \
0      941 Thorpe St  Rock Springs    WY    82901   
1     2809 HARRIS DR       Antioch    CA    94509   
2  1441 Eastlake Ave   Los Angeles    CA    90033   

                          full_address  
0     941 Thorpe StRock SpringsWY82901  
1         2809 HARRIS DRAntiochCA94509  
2  1441 Eastlake AveLos AngelesCA90033

In [272]:
test_ad.head(3)

uid            address          city state    zip  \
0    1      941 Thorpe St  Rock Springs    WY  82901   
1    2     2809 HARRIS DR       Antioch    CA  94509   
2    3  1441 Eastlake Ave   Los Angeles    CA  90033   

                            Cust_address                         full_address  
0     941 Thorpe StRock SpringsWY82901.0     941 Thorpe StRock SpringsWY82901  
1         2809 HARRIS DRAntiochCA94509.0         2809 HARRIS DRAntiochCA94509  
2  1441 Eastlake AveLos AngelesCA90033.0  1441 Eastlake AveLos AngelesCA90033

In [274]:
lst_1 = real_ad['full_address'].to_list()
lst_2 = test_ad['full_address'].to_list()

In [280]:
def common_elements(list1, list2):
    result = []
    for element in list1:
        if element in list2:
            result.append(element)
    return result

In [282]:
match = common_elements(lst_1,lst_2)

In [285]:
match2 = common_elements(lst_2,lst_1)

In [284]:
len(match)

74803

In [286]:
len(match2)

64526

In [287]:
def dff_elements(list1, list2):
    return [element for element in list1 if element not in list2]

In [289]:
diff = dff_elements(lst_2,lst_1)
len(diff)

34723

In [291]:
34723+64526

99249

In [36]:
#examine where address is '--'
real_ad[real_ad.address == '- -'].head()

address          city state    zip
2618     - -  Butte Valley    CA  95965
2986     - -      Palmdale    CA  93552
3037     - -     Lancaster    CA  93536
4010     - -    Santa Rosa    CA  95407
4113     - -       Shandon    CA  93461

In [37]:
#examine address that start with '-'
real_ad[real_ad['address'].astype(str).str.startswith('-')].head()

address          city state    zip
529   - E Ave R 10             -    CA  93543
2091     - 17th St      Piedmont    CA  94611
2618           - -  Butte Valley    CA  95965
2986           - -      Palmdale    CA  93552
3037           - -     Lancaster    CA  93536

In [38]:
#examine address that start with '-'
real_ad[real_ad['address'].astype(str).str.startswith('')].head()

address           city state    zip
0      941 Thorpe St   Rock Springs    WY  82901
1     2809 HARRIS DR        Antioch    CA  94509
2  1441 Eastlake Ave    Los Angeles    CA  90033
3       7 Eucalyptus  Newport Beach    CA  92657
4    1400 Lachman Ln    Los Angeles    CA  90272

In [39]:
#examine address that start with '-'
real_ad[real_ad['address'].astype(str).str.startswith('#')].count()

address    504
city       504
state      504
zip        504
dtype: int64

In [40]:
real_ad[(real_ad['address'].astype(str).str.startswith('#'))].to_csv 
        #&
        #(real_ad['address'] != '#NAME?')].head()

<bound method NDFrame.to_csv of        address         city state    zip
32      #NAME?            -    CA  93551
77      #NAME?  Victorville    CA  92395
106     #NAME?      Hayward    CA      -
536     #NAME?    Lancaster    CA  93535
783     #NAME?     Redlands    CA  92374
...        ...          ...   ...    ...
128581  #NAME?     Petaluma    CA  94954
128861  #NAME?            -    CA  93591
129834  #NAME?   Ridgecrest    CA  93555
129968  #NAME?            -    CA      -
129987  #NAME?     Palmdale    CA  93551

[504 rows x 4 columns]>

In [41]:
real_ad.head()

address           city state    zip
0      941 Thorpe St   Rock Springs    WY  82901
1     2809 HARRIS DR        Antioch    CA  94509
2  1441 Eastlake Ave    Los Angeles    CA  90033
3       7 Eucalyptus  Newport Beach    CA  92657
4    1400 Lachman Ln    Los Angeles    CA  90272

In [42]:
real_ad['address'] = real_ad['address'].str.strip()

In [6]:
cols = ['address', 'city', 'state', 'zip', 'full_address']
real_ad[cols] = real_ad[cols].apply(lambda x: x.str.strip())

In [ ]:
real_ad.head()

In [166]:
real_ad['full_address'] = real_ad.address + real_ad.city+real_ad.state+real_ad.zip
real_ad.head()

address           city state    zip  \
0      941 Thorpe St   Rock Springs    WY  82901   
1     2809 HARRIS DR        Antioch    CA  94509   
2  1441 Eastlake Ave    Los Angeles    CA  90033   
3       7 Eucalyptus  Newport Beach    CA  92657   
4    1400 Lachman Ln    Los Angeles    CA  90272   

                          full_address  
0     941 Thorpe StRock SpringsWY82901  
1         2809 HARRIS DRAntiochCA94509  
2  1441 Eastlake AveLos AngelesCA90033  
3     7 EucalyptusNewport BeachCA92657  
4    1400 Lachman LnLos AngelesCA90272

In [169]:
test_ad['full_address'] = test_ad.address+test_ad.city+test_ad.state+test_ad.zip.astype(str)
test_ad.head()

uid            address           city state    zip  \
0    1      941 Thorpe St   Rock Springs    WY  82901   
1    2     2809 HARRIS DR        Antioch    CA  94509   
2    3  1441 Eastlake Ave    Los Angeles    CA  90033   
3    4         7 ucayptus  Newport Beach    CA  92657   
4    5    1400 Lachman Ln            NaN    CA  90272   

                            Cust_address                         full_address  
0     941 Thorpe StRock SpringsWY82901.0     941 Thorpe StRock SpringsWY82901  
1         2809 HARRIS DRAntiochCA94509.0         2809 HARRIS DRAntiochCA94509  
2  1441 Eastlake AveLos AngelesCA90033.0  1441 Eastlake AveLos AngelesCA90033  
3       7 ucayptusNewport BeachCA92657.0       7 ucayptusNewport BeachCA92657  
4                                    NaN                                  NaN

In [ ]:
test_ad['full_address'] = test_ad.full_address.str.upper()

In [ ]:
test_ad.head()

In [168]:
test_ad['zip'] = test_ad['zip'].astype(str).apply(lambda x: x.replace('.0',''))
test_ad.head()

uid            address           city state    zip  \
0    1      941 Thorpe St   Rock Springs    WY  82901   
1    2     2809 HARRIS DR        Antioch    CA  94509   
2    3  1441 Eastlake Ave    Los Angeles    CA  90033   
3    4         7 ucayptus  Newport Beach    CA  92657   
4    5    1400 Lachman Ln            NaN    CA  90272   

                            Cust_address  \
0     941 Thorpe StRock SpringsWY82901.0   
1         2809 HARRIS DRAntiochCA94509.0   
2  1441 Eastlake AveLos AngelesCA90033.0   
3       7 ucayptusNewport BeachCA92657.0   
4                                    NaN   

                            full_address  
0     941 Thorpe StRock SpringsWY82901.0  
1         2809 HARRIS DRAntiochCA94509.0  
2  1441 Eastlake AveLos AngelesCA90033.0  
3       7 ucayptusNewport BeachCA92657.0  
4                                    NaN

In [177]:
new_df = pd.concat([real_ad,test_ad],sort = True)
new_df.head()

Cust_address            address           city  \
0          NaN      941 Thorpe St   Rock Springs   
1          NaN     2809 HARRIS DR        Antioch   
2          NaN  1441 Eastlake Ave    Los Angeles   
3          NaN       7 Eucalyptus  Newport Beach   
4          NaN    1400 Lachman Ln    Los Angeles   

                          full_address state  uid    zip  
0     941 Thorpe StRock SpringsWY82901    WY  NaN  82901  
1         2809 HARRIS DRAntiochCA94509    CA  NaN  94509  
2  1441 Eastlake AveLos AngelesCA90033    CA  NaN  90033  
3     7 EucalyptusNewport BeachCA92657    CA  NaN  92657  
4    1400 Lachman LnLos AngelesCA90272    CA  NaN  90272

In [178]:
new_df.isna().sum()

Cust_address    138715
address            295
city              1298
full_address      9010
state             7449
uid             130000
zip                 16
dtype: int64

In [181]:
new_df[new_df.full_address.isna()].head()

Cust_address address           city full_address state  uid    zip
341           NaN     NaN       Millbrae          NaN    CA  NaN  94030
764           NaN     NaN      Sunnyvale          NaN    CA  NaN  94086
967           NaN     NaN       Hesperia          NaN    CA  NaN  92345
1034          NaN     NaN     Sacramento          NaN    CA  NaN  95864
1057          NaN     NaN  San Francisco          NaN    CA  NaN  94124

In [190]:
new_df[new_df.state.isna()].head(3)

Cust_address address city full_address state  uid  zip
14011          NaN     NaN  NaN          NaN   NaN  NaN  NaN
23526          NaN     NaN  NaN          NaN   NaN  NaN  NaN
27995          NaN     NaN  NaN          NaN   NaN  NaN  NaN

In [189]:
new_df[(new_df.city == 'Lancaster') & (new_df.address == '1830 EINSTEIN ST')]

Cust_address           address       city  \
50264           NaN  1830 EINSTEIN ST  Lancaster   
115282          NaN  1830 EINSTEIN ST  Lancaster   
49905           NaN  1830 EINSTEIN ST  Lancaster   

                            full_address state      uid    zip  
50264   1830 EINSTEIN STLancasterCA93535    CA      NaN  93535  
115282  1830 EINSTEIN STLancasterCA93535    CA      NaN  93535  
49905                                NaN   NaN  50265.0  93535

In [ ]:
test_ad.dtypes

In [ ]:
test_ad.isna().sum()

In [ ]:
incomplete = test_ad[test_ad['full_address'].isna()]
incomplete.head()

In [ ]:
na_vals = ['- - ','np.nan','#NAME?','#NAM?']

In [ ]:
real_ad

- With complete address:
    - Examine patterns (if any) where certain columns are missing 
        - For example: missing street name, city, state, zip, etc

In [ ]:
incomplete[incomplete.isin(na_vals)].head()

In [ ]:
test_ad.full_address.describe()

In [ ]:
real_ad.full_address.describe()

In [ ]:
#Rename columns in test ad 
#Join test_ad to real_ad  

In [ ]:
df1 = real_ad[['full_address']]
df2 = test_ad[['full_address']]

In [ ]:
df1.merge(df2, left_on='full_address', right_on='address')

In [ ]:
df3 = df1.merge(df2, left_on='full_address', right_on='address')

In [ ]:
df3.loc[df3.duplicated(keep='first'), :]

In [ ]:
# examine the duplicate rows (including all duplicates)
dupes = df3.loc[df3.duplicated(keep=False), :]

In [ ]:
dupes.shape

In [ ]:
df1.shape

In [ ]:
df3.query('address == full_address')

In [2]:
import pandas as pd
import numpy as np

In [9]:
test_ad.state.unique()

array(['WY', 'CA', 'NY', 'TX', 'TN', 'PA', 'MI', 'FL', 'GA', 'MO', 'AR',
       'CO', 'WA', 'OH', 'AZ', 'MA', 'IL', 'CT', 'SC', 'KY', 'NV', 'MN',
       'AL', 'NC', 'OK', 'WV', 'VA', 'UT', 'VT', 'ID', 'NJ', 'IN', 'MD',
       'SD', 'OR', 'NH', 'NM', 'IA', 'KS', 'LA', 'WI', 'ME', 'NE', 'MS',
       'MT', 'DE', 'ND', 'ca', 'ma', 'ia', 'nd', 'oh', 'id', 'il', 'HI',
       'tx', 'ut', 'pa', 'sd', 'ny', 'RI', 'co', 'wa', 'va', 'mn', 'az',
       'nc', 'tn', nan, 'ne', 'ky', 'nj', 'la', 'ct', 'ks', 'mo', 'md',
       'al', 'wi', 'mi', 'wy', 'AK', 'fl', 'in', 'DC', 'ri', 'or', 'nh',
       'hi', 'nm', 'mt', 'ms', 'ga', 'nv', 'ar', 'ok', 'me'], dtype=object)

In [10]:
test_ad.state.nunique()

96